neqo-transport/src/fc.rs

```diff
@@ -904,4 +1017,215 @@ mod test {
+        let initial_max_active = fc.max_active();
+
+        // Consume and retire multiple receive windows without increasing time.
+        for _ in 1..11 {
+            let consumed = fc.set_consumed(fc.next_limit())?;
+            fc.add_retired(consumed);
+            write_frames(&mut fc, rtt, now);
+        }
+        let increased_max_active = fc.max_active();
+
+        assert!(
+            initial_max_active < increased_max_active,
+            "expect receive window auto-tuning to increase max_active on 0 latency connection"
+        );
+
+        // Huge idle time.
+        now += Duration::from_secs(60 * 60); // 1h
+        let consumed = fc.set_consumed(fc.next_limit()).unwrap();
+        fc.add_retired(consumed);
+
+        assert_eq!(write_frames(&mut fc, rtt, now), 1);
+        assert_eq!(
+            increased_max_active,
+            fc.max_active(),
+            "expect receive window auto-tuning never to decrease max_active"
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn stream_data_blocked_triggers_auto_tuning() -> Res<()> {
+        let rtt = Duration::from_millis(40);
+        let now = Instant::now();
+        let mut fc = ReceiverFlowControl::new(StreamId::new(0), INITIAL_RECV_WINDOW_SIZE as u64);
+
+        // Send first window update to give auto-tuning algorithm a baseline.
+        let consumed = fc.set_consumed(fc.next_limit())?;
+        fc.add_retired(consumed);
+        assert_eq!(write_frames(&mut fc, rtt, now), 1);
+
+        // Use up a single byte only, i.e. way below WINDOW_UPDATE_FRACTION.
+        let consumed = fc.set_consumed(fc.retired + 1)?;
+        fc.add_retired(consumed);
+        assert_eq!(
+            write_frames(&mut fc, rtt, now),
+            0,
+            "expect receiver to not send window update unprompted"
+        );
+
+        // Receive STREAM_DATA_BLOCKED frame.
+        fc.send_flowc_update();
+        let previous_max_active = fc.max_active();
+        assert_eq!(
+            write_frames(&mut fc, rtt, now),
+            1,
+            "expect receiver to send window update"
+        );
+        assert!(
+            previous_max_active < fc.max_active(),
+            "expect receiver to auto-tune (i.e. increase) max_active"
+        );
+
+        Ok(())
+    }
+
+    #[expect(clippy::cast_precision_loss, reason = "This is test code.")]
+    #[test]
+    fn auto_tuning_approximates_bandwidth_delay_product() -> Res<()> {
+        const DATA_FRAME_SIZE: u64 = 1_500;
+        /// Allow auto-tuning algorithm to be off from actual bandwidth-delay
+        /// product by up to 1KiB.
+        const TOLERANCE: u64 = 1024;
+
+        test_fixture::fixture_init();
+
+        // Run multiple iterations with randomized bandwidth and rtt.
+        for _ in 0..1_000 {
+            // Random bandwidth between 1 Mbit/s and 1 Gbit/s.
+            let bandwidth = u64::from(u16::from_be_bytes(random::<2>()) % 1_000 + 1) * 1024 * 1024;
```

Comment:
```suggestion
            let bandwidth = u64::from(u16::from_be_bytes(random::<2>()) % 1_000 + 1) * 1_000 * 1_000;
```

I'm curious as to why 1_000 is needed but 1024 is accepted.